### PR TITLE
Issue 110: Add epiweekly fits and scoring

### LINKED ISSUE #110: Scoring on (epi)weekly scale
At the moment scoring (on the scoring branch) is done daily, and averaged over a week.

This could be useful to keep this, but noting that this includes scoring within-week effects, but we should also have weekly aggregated then scored utility.

### Sub-task planning

- [x] Land #148 
- [x] Add generate epiweekly truth data from daily truth data step
- [x] Add run baseline models on epiweekly
- [x] Add extra pyrenew-hew postprocess to chunk the daily posterior preds into epiweekly
- [x] check how much revision the scoring function needs (might be zero)

### PR DESCRIPTION
**Pull request by SamuelBrand1**
This PR closes #110 .

NB this PR uses `forecasttools` functions, in particular `daily_to_epiweekly`. Very handy.

The goal is to compare the _epiweekly_ forecasts of baseline models (which are fitted to epiweekly data) to the _epiweekly_ grouped forecasts from `pyrenew_hew`.

The daily analysis is unaffected; that is baseline models are still run for daily reporting as well.

## Tasks to contribute
- [x] Add generate epiweekly truth data from daily truth data step
- [x] Add run baseline models on epiweekly
- [x] Add extra pyrenew-hew postprocess to chunk the daily posterior preds into epiweekly
- [x] check how much revision the scoring function needs (might be zero)

## Testing
At each step the test mode pipeline is run to check for failure or breaking changes.

### CODE REVIEW

**Comment by SamuelBrand1** [comment]
Added a step to convert daily resolved data to epiweekly cf `generate_epiweekly.R`.

**Comment by SamuelBrand1** [comment]
Modified `timeseries_forecasts.R` to run baseline forecasters for both for daily and epiweekly models.

**Comment by SamuelBrand1** [comment]
Added scoring on epiweekly scale for `pyrenew`. The algorithm is that if `epiweekly = TRUE` then the pyrenew forecasts for counts and other counts are group summed by epiweek (which auto does by draw), pivoted out to create the correct _weekly_ percentage (which is not the mean of the daily percentages), and then pivoted longer.

Dates for the week are supplied by more forecasttools helpers.

No other changes were need because the scoring is flexible enough to ingest this.

File: pipelines/score_forecast.R
Comment (damonbayer) [inline pipelines/score_forecast.R:162]:
```suggestion
  if (epiweekly) {
    eval_data_filename <- glue::glue("epiweekly_{eval_data_filename}")
  }
```

**Comment by damonbayer** [comment]
@SamuelBrand1 Since we are making even more model results and anticipate adding PyRenew-EW to the workflow soon, I would really like to push toward https://github.com/CDCgov/pyrenew-hew/issues/101. Could the fable-related parts of that task be accomplished in this PR (i.e. let's store all the results from these time series models in a `fable` directory)?

File: pipelines/score_forecast.R
Comment (damonbayer) [inline pipelines/score_forecast.R:207]: Should something like this instead be done in `pipelines/postprocess_state_forecast.R`?

We would actually like to create epiweekly forecasts and present them, right? See https://github.com/CDCgov/pyrenew-hew/issues/79

Comment (dylanhmorris) [reply pipelines/score_forecast.R:207]: Yes, I think it makes sense to have an epiweekly forecast as canonical output, not just for scoring.

Related: perhaps the location-specific runs should build a location-specific hubverse table, and the overall hubverse table can then just be built via collation (right now it's built via iteration over the results in `create_hubverse_table.R`)

Comment (dylanhmorris) [reply pipelines/score_forecast.R]: I had need of this just now so I did it in 6627518

**Comment by damonbayer** [comment]
> > @SamuelBrand1 Since we are making even more model results and anticipate adding PyRenew-EW to the workflow soon, I would really like to push toward #101. Could the fable-related parts of that issue be accomplished in this PR (i.e. let's store all the results from these time series models in a `fable` directory)?
> 
> I'd rather avoid mission creep on PRs and complete then do another issue.

I don't see the point of adding new code that saves things in locations that we don't like, but I defer to you.

File: pipelines/batch/setup_prod_job.py
Comment (damonbayer) [inline pipelines/batch/setup_prod_job.py:2]:
```suggestion
Set up a multi-location, multi-disease production run
```

File: pipelines/batch/setup_eval_job.py
Comment (damonbayer) [inline pipelines/batch/setup_eval_job.py:25]: Why is the output directory not configurable? And if it were, shouldn't we just have one arg where you can put the full path?

File: pipelines/batch/setup_eval_job.py
Comment (damonbayer) [inline pipelines/batch/setup_eval_job.py:178]: Should the "output" part be configurable?

File: pipelines/batch/setup_parameter_inference_job.py
Comment (damonbayer) [inline pipelines/batch/setup_parameter_inference_job.py:142]: What are these?

File: pipelines/batch/setup_parameter_inference_job.py
Comment (damonbayer) [inline pipelines/batch/setup_parameter_inference_job.py:1]: What is this job for? Seems like it uses more data and different priors than the usual run?

File: pipelines/batch/setup_prod_job.py
Comment (damonbayer) [inline pipelines/batch/setup_prod_job.py:177]: I have the same concerns about output here as in `setup_eval_job.py`

File: pipelines/timeseries_forecasts.R
Comment (damonbayer) [inline pipelines/timeseries_forecasts.R:159]: I think either
```r
if (epiweekly) {
  # Assign variables
} else {
  # Assign variables
}
```
or separate
```r
prefix <- if_else(epiweekly, "epiweekly_", "")
```
is easier to read

File: pipelines/score_hubverse.R
Comment (damonbayer) [inline pipelines/score_hubverse.R:55]: can these assigned programmatically (`dir_ls`?)

File: pipelines/score_hubverse.R
Comment (damonbayer) [inline pipelines/score_hubverse.R:62]: why?

File: pipelines/score_hubverse.R
Comment (damonbayer) [inline pipelines/score_hubverse.R:68]: why?

File: pipelines/score_hubverse.R
Comment (damonbayer) [inline pipelines/score_hubverse.R:79]: Overall seems like you could `walk` or loop though each disease and use a common function.

File: pipelines/score_hubverse.R
Comment (damonbayer) [inline pipelines/score_hubverse.R:85]: What do you do with this figure? Should it be saved somewhere?

File: pipelines/create_observed_data_tables.py
Comment (damonbayer) [inline pipelines/create_observed_data_tables.py:1]: General comment that the pipelines folder is becoming unwieldy. After this PR, I count 21 scripts. Additional organization would be helpful.

File: pipelines/default_priors.py
Comment (damonbayer) [inline pipelines/default_priors.py:1]: Why is the filemode changing?

Comment (damonbayer) [inline pipelines/default_priors.py:1]: What is motivating the changes to the priors?

File: pipelines/generate_epiweekly.R
Comment (damonbayer) [inline pipelines/generate_epiweekly.R:33]: I think it makes more sense to take the `data_path` as the arg and use utilities from `fs` to extract the parts as necessary.

File: pipelines/plot_parameters.R
Comment (damonbayer) [inline pipelines/plot_parameters.R:1]: Does this plot get saved somewhere?

**Review by damonbayer — COMMENTED** [review]
Nothing that would necessarily prevent a merge, but I had lots of comments and questions.

File: pipelines/batch/setup_eval_job.py
Comment (dylanhmorris) [reply pipelines/batch/setup_eval_job.py:178]: No, because that's the hard-coded bind mount location of the output blob storage container within the running podman/docker container.

File: pipelines/batch/setup_eval_job.py
Comment (dylanhmorris) [inline pipelines/batch/setup_eval_job.py:119]: Noting the target of `output`

File: pipelines/batch/setup_eval_job.py
Comment (dylanhmorris) [reply pipelines/batch/setup_eval_job.py:178]: See https://github.com/CDCgov/pyrenew-hew/pull/159/files/a2fa9afa9651fda572c50b4a2db90b8379dc468b#r1859432678

File: pipelines/batch/setup_eval_job.py
Comment (dylanhmorris) [reply pipelines/batch/setup_eval_job.py:25]: See https://github.com/CDCgov/pyrenew-hew/pull/159/files/a2fa9afa9651fda572c50b4a2db90b8379dc468b

Because this is just an alias for the root of the output blob storage container, I don't think it needs a configurable name.

File: pipelines/batch/setup_parameter_inference_job.py
Comment (dylanhmorris) [reply pipelines/batch/setup_parameter_inference_job.py:1]: Yes, idea is to use this script for runs with longer timeseries and less informative priors. This might be useful, e.g., for choosing relatively informative priors for shorter timeseries fits.

File: pipelines/batch/setup_parameter_inference_job.py
Comment (dylanhmorris) [reply pipelines/batch/setup_parameter_inference_job.py:142]: They live in the `pyrenew-hew-config` blob storage container.

File: pipelines/batch/setup_prod_job.py
Comment (dylanhmorris) [reply pipelines/batch/setup_prod_job.py:177]: See response above.

File: pipelines/create_observed_data_tables.py
Comment (dylanhmorris) [reply pipelines/create_observed_data_tables.py:1]: Agree but out of scope here. And this one is a rename rather than a new file.

File: pipelines/default_priors.py
Comment (dylanhmorris) [reply pipelines/default_priors.py:1]: Guessing this or something like it? https://stackoverflow.com/questions/1257592/how-do-i-remove-files-saying-old-mode-100755-new-mode-100644-from-unstaged-cha

File: pipelines/batch/setup_parameter_inference_job.py
Comment (damonbayer) [reply pipelines/batch/setup_parameter_inference_job.py:142]: Should they be reviewed as part of code review?

File: pipelines/score_hubverse.R
Comment (damonbayer) [reply pipelines/score_hubverse.R:62]: Should the reference date be computed based on the data, rather than the time the script is run?

File: pipelines/timeseries_forecasts.R
Comment (SamuelBrand1) [reply pipelines/timeseries_forecasts.R]: Agreed. Much nicer. I've changed this now.

File: pipelines/batch/setup_eval_job.py
Comment (damonbayer) [reply pipelines/batch/setup_eval_job.py:119]: Why are you noting the target of `output`? Are you uncertain about it?

**Review by damonbayer — APPROVED** [review]
@dylanhmorris This PR has my tentative approval. Please respond to (or make changes based on) the remaining comments before merging.

File: pipelines/generate_epiweekly.R
Comment (SamuelBrand1) [reply pipelines/generate_epiweekly.R]: I left `dataname` as an arg, because `model_run_dir` is a common arg across multiple processes (although obvs we could change this in the associated `main` function so feel free to come back at me on that). 

However, I agree than the `ext` doesn't need to be a separate arg!

File: pipelines/batch/setup_parameter_inference_job.py
Comment (dylanhmorris) [reply pipelines/batch/setup_parameter_inference_job.py:142]: This is a good question about practice for setting configuration and how much this repo is implementation versus configuration, but imo out of scope for this PR. Will open an issue.

File: pipelines/score_hubverse.R
Comment (dylanhmorris) [reply pipelines/score_hubverse.R]: This file shouldn't really be in this PR at all. Going to remove from this PR and merge it to separately, cleaned up and with your (good) comments addressed

Comment (dylanhmorris) [reply pipelines/score_hubverse.R]: This file shouldn't really be in this PR at all. Going to remove from this PR and merge it to separately, cleaned up and with your comments addressed

File: pipelines/batch/setup_eval_job.py
Comment (dylanhmorris) [reply pipelines/batch/setup_eval_job.py:119]: So I could link to it from the other comments.

**Pull request merged**
